### PR TITLE
RSDK-7756: Add SetRPM to gpiostepper and roboclaw motor drivers

### PR DESCRIPTION
**Pull request by martha-johnston**
- need to test on gpiostepper hardware
- roboclaw hardware tests done
- roboclaw doesn't seem to have a test file? so didn't add roboclaw SetRPM software test

### CODE REVIEW

File: components/motor/roboclaw/roboclaw.go [hunk 1]
```diff
@@ -318,7 +319,40 @@ func (m *roboclawMotor) GoTo(ctx context.Context, rpm, positionRevolutions float
...
 func (m *roboclawMotor) SetRPM(ctx context.Context, rpm float64, extra map[string]interface{}) error {
-	return motor.NewSetRPMUnsupportedError(m.Name().ShortName())
+	if m.conf.TicksPerRotation == 0 {
+		if rpm > maxRPM {
+			rpm = maxRPM
+		} else if rpm < -1*maxRPM {
+			rpm = -1 * maxRPM
+		}
+		powerPct := rpm / maxRPM
+		m.logger.CInfo(ctx, "speed is an estimation based on a max RPM 250. For increased accuracy, connect encoders")
```
[the remark this answers is not in the document]
Comment (penguinland) [reply components/motor/roboclaw/roboclaw.go:329]: This line implies that encoders are not already connected, but it's not clear to me that we've ever checked for them. Maybe `TicksPerRotation` is 0 if and only if there are no encoders connected? If so, I'd put a comment about that on line 322. If not, I'd rephrase this line so people who already have encoders connected don't get confused by the message.

File: components/motor/roboclaw/roboclaw.go [hunk 2]
```diff
@@ -318,7 +319,40 @@ func (m *roboclawMotor) GoTo(ctx context.Context, rpm, positionRevolutions float
...
+		err = m.conn.SpeedDistanceM1(m.addr, ticksPerSecond, uint32(math.Inf(int(rpm))), true)
+	case 2:
+		err = m.conn.SpeedDistanceM2(m.addr, ticksPerSecond, uint32(math.Inf(int(rpm))), true)
+	default:
+		return m.conf.wrongChannelError()
+	}
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (penguinland) [inline components/motor/roboclaw/roboclaw.go:355]: To summarize these 4 lines: if the error is not nil, return the error, and if it is nil, return nil (equivalent to returning the error). Why don't we just put return statements on lines 346 and 348, similar to the one on line 350, and get rid of this part? It seems unnecessarily complicated, but perhaps you've got a plan to use the extra complexity in the near future and it'll make more sense then.

**Review by penguinland — APPROVED** [review]
All my suggestions are about adding comments. The code itself LGTM!

File: components/motor/gpiostepper/gpiostepper.go [hunk 3]
```diff
@@ -402,7 +402,25 @@ func (m *gpioStepper) GoTo(ctx context.Context, rpm, positionRevolutions float64
 
 // SetRPM instructs the motor to move at the specified RPM indefinitely.
 func (m *gpioStepper) SetRPM(ctx context.Context, rpm float64, extra map[string]interface{}) error {
-	return motor.NewSetRPMUnsupportedError(m.Name().ShortName())
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	// calculate delay between steps for the thread in the goroutine that we started in component creation
+	m.stepperDelay = time.Duration(int64(float64(time.Minute) / (math.Abs(rpm) * float64(m.stepsPerRotation))))
```
[the remark this answers is not in the document]
Comment (martha-johnston) [reply components/motor/gpiostepper/gpiostepper.go:409]: I'll def add a comment

File: components/motor/gpiostepper/gpiostepper_test.go [hunk 4]
```diff
@@ -510,5 +510,38 @@ func TestRunning(t *testing.T) {
...
+		test.That(t, err, test.ShouldBeNil)
+
+		on, _, err := m.IsPowered(ctx, nil)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, on, test.ShouldEqual, false)
+
+		pos, err := m.Position(ctx, nil)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, pos, test.ShouldBeGreaterThan, 4)
+		test.That(t, pos, test.ShouldBeLessThan, 202)
```
[the remark this answers is not in the document]
Comment (martha-johnston) [reply components/motor/gpiostepper/gpiostepper_test.go:543]: using the same values as other tests, unsure what the logic was for choosing them

File: components/motor/roboclaw/roboclaw.go [hunk 5]
```diff
@@ -273,6 +273,7 @@ func (m *roboclawMotor) GoFor(ctx context.Context, rpm, revolutions float64, ext
 			return errors.Wrap(err, "error in GoFor")
 		}
 		if revolutions == 0 {
+			m.logger.CWarn(ctx, "Deprecated: setting revolutions == 0 will spin the motor indefinitely at the specified RPM")
```
[the remark this answers is not in the document]
Comment (martha-johnston) [reply components/motor/roboclaw/roboclaw.go:276]: ticket to remove it on 9/1!

File: components/motor/roboclaw/roboclaw.go (hunk 2 above)
Comment (martha-johnston) [reply components/motor/roboclaw/roboclaw.go:355]: yeah good point. this is copied from another section of code so I will fix it in both

File: components/motor/gpiostepper/gpiostepper_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (penguinland) [reply components/motor/gpiostepper/gpiostepper_test.go:543]: :shrug: SGTM

File: components/motor/gpiostepper/gpiostepper.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (randhid) [reply components/motor/gpiostepper/gpiostepper.go:409]: I think we can do
```suggestion
	m.stepperDelay = time.Duration(time.Minute / (int64(math.Abs(rpm) * m.stepsPerRotation))
```
We need to cast to a integer anyway since the stepper steps are discrete, and it simplifies this line a little.
Also this is prime for abstracting out into testable math to avoid future breakages, like the math functions at the core of the gpio motor driver.

[the remark this answers is not in the document]
Comment (penguinland) [reply components/motor/gpiostepper/gpiostepper.go:409]: That approach would lead to roundoff errors. For example, if stepsPerRotation is 1 and rpm is 1.5, the stepper delay would be 1 minute with Rand's approach, but it ought to be 0.67 minutes instead. Casting to int should be the last thing done, to retain precision.

[the remark this answers is not in the document]
Comment (randhid) [reply components/motor/gpiostepper/gpiostepper.go:409]: I guess I'm confused because I don't think we can pass a non-nanosecond time magnitude into time.Duration, and the other two values are integers, aren't we only losing decimal precision on rpm?

[the remark this answers is not in the document]
Comment (penguinland) [reply components/motor/gpiostepper/gpiostepper.go:409]: Ohh, if `rpm` and `stepsPerRotation` are both integers already, you're right that there's no need to do any casting anywhere (maybe from int32 to int64). I had assumed we were casting certain parts to floats because other parts were floats already.

[the remark this answers is not in the document]
Comment (randhid) [reply components/motor/gpiostepper/gpiostepper.go:409]: Just to be clear `time.Minute` [value set at 60e9 nanosecond] and `m.stepsPerRotation` are integers, `rpm` is a float64, we're losing decimal precision on rpm

But this math should be equivalent as well and might be clearer
```go
time.Duration(
          int64 (
          float64(time.Minute /  m.stepsPerRotation) / math.Abs(rpm)
          )
```

[the remark this answers is not in the document]
Comment (randhid) [reply components/motor/gpiostepper/gpiostepper.go:409]: We have our greatest calculation error here when the delay we calculate is on the order of nanoseconds. We protect against that with the minimum stepper delay, and stepper motors driven by gpio drivers do not really move themselves on a per nanosecond basis 😄

[the remark this answers is not in the document]
Comment (randhid) [reply components/motor/gpiostepper/gpiostepper.go:409]: Actually, never mind, I mistakenly thought time.Minute would output an integer, it outputs a Duration, so we need to cast it to a int or float anyway.

[the remark this answers is not in the document]
Comment (randhid) [reply components/motor/gpiostepper/gpiostepper.go:409]: Regardless @martha-johnston my main ask for this line is to take this particular line out and test the output anyway.

File: components/motor/roboclaw/roboclaw.go [hunk 6]
```diff
@@ -318,7 +319,40 @@ func (m *roboclawMotor) GoTo(ctx context.Context, rpm, positionRevolutions float
 }
 
 func (m *roboclawMotor) SetRPM(ctx context.Context, rpm float64, extra map[string]interface{}) error {
-	return motor.NewSetRPMUnsupportedError(m.Name().ShortName())
+	if m.conf.TicksPerRotation == 0 {
```
Comment (randhid) [inline components/motor/roboclaw/roboclaw.go:322]: There a huge bug in this driver, where if the ticks per rotation is zero, this motor is actually not position reporting, please make a follow up ticket.

Once we have a valid Proeprties function we can call that to check what power percentage to calculate in each case within this driver.

File: components/motor/gpiostepper/gpiostepper.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (martha-johnston) [reply components/motor/gpiostepper/gpiostepper.go:409]: okay. I'll write a software test and test in person once I get the gpiostepper working

**Comment by martha-johnston** [comment]
only thing I haven't touched yet is the stepper delay, want to remove the line and test on hardware as rand requested

**Comment by martha-johnston** [comment]
@penguinland @randhid I tried some different variations of the stepper delay calculation with hardware tests and I think this is the best way to calculate it without it being truncated to 0 with all the int and time casting. I added one more comment that hopefully clarifies what the calculation is doing.

**Comment by randhid** [comment]
> @penguinland @randhid I tried some different variations of the stepper delay calculation with hardware tests and I think this is the best way to calculate it without it being truncated to 0 with all the int and time casting. I added one more comment that hopefully clarifies what the calculation is doing.

Please remember to make it a testable function and use that same function in the two places where the calculation is used.

**Comment by martha-johnston** [comment]
if either of you have a minute could you take a look at the failing test? it's not clear to me what is failing I must be missing the test name
